src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptions.java

```diff
@@ -596,6 +588,90 @@ public boolean apply(Object input) {
     }
 
     /**
+     * Tests that emulates a frozen elected master node that unfreezes and pushes his cluster state to other nodes
+     * that already are following another elected master node. These nodes should reject this cluster state and prevent
+     * them from following the stale master.
+     */
+    @Test
+    public void testStaleMasterNotHijackingMajority() throws Exception {
+        // TODO: on mac OS multicast threads are shared between nodes and we therefore we can't simulate GC and stop pinging for just one node
+        // find a way to block thread creation in the generic thread pool to avoid this.
+        // 3 node cluster with unicast discovery and minimum_master_nodes set to 2:
+        List<String> nodes = startUnicastCluster(3, null, 2);
+
+        // Save the current master node as old master node, because that node will get frozen
+        final String oldMasterNode = internalCluster().getMasterName();
+        for (String node : nodes) {
+            client(node).admin().cluster().prepareHealth().setWaitForEvents(Priority.LANGUID).setLocal(true).get();
+            ensureStableCluster(3, node);
+        }
+        assertMaster(oldMasterNode, nodes);
+
+        // Simulating a painful gc by suspending all threads for a long time on the current elected master node.
+        SingleNodeDisruption masterNodeDisruption = new LongGCDisruption(getRandom(), oldMasterNode);
+
+        // Save the majority side
+        final List<String> majoritySide = new ArrayList<>(nodes);
+        majoritySide.remove(oldMasterNode);
+
+        // Keep track of the masters that appear in the cluster state on both nodes on the majority side,
+        // only new_master may appear in here.
+        final Set<String> masters = new HashSet<>();
+        for (String node : majoritySide) {
+            internalCluster().getInstance(ClusterService.class, node).add(new ClusterStateListener() {
+                @Override
+                public void clusterChanged(ClusterChangedEvent event) {
+                    DiscoveryNode masterNode = event.state().getNodes().masterNode();
```

Comment:
can we track changes here instead of master nodes? I think it would be interesting to test we don't have  newMaster -> null -> newMaster elected again situation. Something ala (pseudo code :)):

```
final Map<String,List<Tuple<String,String>>> masterChangePerNode
```
and then:

```
if (event.previousState.masterNodeId != event.state.masterNodeId) { 
  synchronize(masterChangePerNode) {
      masterChangePerNode.addToNodeId(new Tuple<oldMasterId,newMasterId>)
 }
```